## Python进阶: 异常处理

-Author: bavdu

-Email: bavduer@163.com

-Github: https://github.com/bavdu

---



实际的工作过程中, 有很多错误情况出现:

**场景:** 工程师提交了代码, 不过代码某处忘记了异常处理. 碰巧这种异常发生的频率还不低, 所以在代码push到线上之后不久, 就会收到紧急通知—服务器崩溃了



####错误 与 异常

---

首先需要了解的是python中的错误和异常是什么？两者之间又有什么联系和区别呢？

**程序中的错误包含两种: 语法错误 和 异常**

- 语法错误: 就是你写的代码不符合编程规范, 无法被识别和执行(比如: 书写的过程中漏掉了逗号、冒号等)

  ```python
  if name is not None
      print(name)
    
  ##########################################Out:##########################################
      if name is not None
                        ^
  SyntaxError: invalid syntax
  ```

- 异常: 指的是程序语法结构没有问题, 也可以被执行. 但是在执行的过程中遇到了错误, 这时抛出的就是异常了

  ```python
  10 / 0
  ##########################################Out:##########################################
  Traceback (most recent call last):
    File "<stdin>", line 1, in <module>
  ZeroDivisionError: integer division or modulo by zero
  
  order * 2
  ##########################################Out:##########################################
  Traceback (most recent call last):
    File "<stdin>", line 1, in <module>
  NameError: name 'order' is not defined
  
  1 + [1, 2]
  ##########################################Out:##########################################
  Traceback (most recent call last):
    File "<stdin>", line 1, in <module>
  TypeError: unsupported operand type(s) for +: 'int' and 'list'
  
  ```

  

#### 处理异常

---

程序执行到程序中某处抛出了异常, 程序就会被终止并退出. 那么有没有办法可以不终止程序, 让其照样运行下去呢？答案是肯定的! 这就需要用到我们的异常处理机制了, 通常使用try…except…finally…来解决

```python
try:
  	file = open('file.txt', 'r')
    content = file.read()
    print(content)
except Exception as error:		## Exception是其他所有非系统异常的基类,能够匹配任意非系统异常
    print('[ERROR] {}'.format(error))
finally:
    print('program end')

print('continue running...')
number = [1, 2, 3, 4, 5]
for element in number:
    if element % 2 == 0:
        print(element)
```

拓展except处理机制

```python
try:
    s = input('please enter two numbers: ')
    num1 = int(s.split(',')[0].strip())
    num2 = int(s.split(',')[1].strip())
except (ValueError, IndexError) as error:   ## 类似if 为True 则不判断其他的
    print('Error: {}'.format(error))
    
print('continue...')
```

```python
try:
    s = input('please enter two numbers: ')
    num1 = int(s.split(',')[0].strip())
    num2 = int(s.split(',')[1].strip())
except ValueError as error:
    print('Value Error: {}'.format(error))
except IndexError as error:
    print('Index Error: {}'.format(error))

print('continue...')
```

省略异常类型不写, 就是匹配所有的异常(包括系统异常)

```python
try:
    s = input('please enter two numbers: ')
    num1 = int(s.split(',')[0].strip())
    num2 = int(s.split(',')[1].strip())
except ValueError as error:
    print('Value Error: {}'.format(error))
except IndexError as error:
    print('Index Error: {}'.format(error))
except:          ##无内容，捕捉所有异常
    print('Other error')

print('continue...')
```



#### 自定义异常抛出

---

有时在程序中我们也可以自定义异常的抛出, python中有个关键字: raise, 可以抛出异常

```python
class MyInputError(Exception):
    def __init__(self, value):
        self.value = value

    def __str__(self):
        return '{} is invalid input'.format(self.value)


try:
    string = input('>>>Input: ')
    if 'abcd' in string:
        raise MyInputError(string)
    else:
        print(string)
except MyInputError as error:
    print(error)
```



#### 异常的使用场景 和 注意点

---

- 注意点: 不能滥用异常处理

  ```python
  d = {'name': 'jason', 'age': 20}
  try:
      value = d['dob']
  except KeyError as err:
      print('KeyError: {}'.format(err))
  ```

- 场景: 在运维的过程中, 我们通常会使用python做服务器信息的数据报表(比如: 某业务的CPU、MEM、DISK的使用情况), 使用情况通常都是以key-value的形式存在于数据库中的, 当需要时可以根据CPU、MEM、DISK查询数据库中的记录, 就能返回结果.

  而数据库放回的原始数据, 往往是JSON String的形式, 那么我们就需要对数据进行解码, 通常是下面这种形式: 

  ```python
  import pymysql
  import json
  
  
  def queryDB(uid):
      def load(user='bavduer', password='(Bavduer..0712)'):
          connection = pymysql.connect(host='39.100.110.135',
                                       user='bavduer',
                                       password='(Bavduer..0712)',
                                       db='pyOpera')
          return connection
      info = {'email': None, 'password': None}
      with load().cursor() as cursor:
          sql = "select email, password from pyOpera.users where id = {}"
          cursor.execute(sql.format(uid))
          result = cursor.fetchall()
          info['email'] = result[0][0]
          info['password'] = result[0][1]
      return json.dumps(info)
  
  
  
  try:
      rawData = queryDB(uid=1)
      data = json.loads(rawData)
      print(data, type(data))
  except Exception as error:
      print('JSONDecodeError: {}'.format(error))
  ```

  